tests/test_bidstream_client.py

```diff
@@ -0,0 +1,160 @@
+                                                                                  99999, 86400,
+                                                                                  max_bidstream_lifetime_seconds=max_bidstream_lifetime)
+                self._client.refresh_keys()
+                with self.assertRaises(EncryptionError) as context:
+                    self._client.decrypt_token_into_raw_uid(token, None)
+                self.assertEqual('invalid token lifetime', str(context.exception))
+
+    def test_token_generated_in_the_future_to_simulate_clock_skew(self, mock_refresh_bidstream_keys):  # TokenGeneratedInTheFutureToSimulateClockSkew
+        created_at_future = dt.datetime.now(tz=timezone.utc) + dt.timedelta(minutes=31)  #max allowed clock skew is 30m
+        for expected_scope, expected_version in test_cases_all_scopes_all_versions:
+            with self.subTest(expected_scope=expected_scope, expected_version=expected_version):
+                token = generate_uid_token(expected_scope, expected_version, created_at=created_at_future)
+                mock_refresh_bidstream_keys.return_value = EncryptionKeysCollection([master_key, site_key],
+                                                                                  expected_scope, site_id, 1,
+                                                99999, 86400)
+                self._client.refresh_keys()
+                with self.assertRaises(EncryptionError) as context:
+                    self._client.decrypt_token_into_raw_uid(token, None)
+                self.assertEqual('invalid token lifetime', str(context.exception))
+
+    def test_token_generated_in_the_future_within_allowed_clock_skew(self, mock_refresh_bidstream_keys):  # TokenGeneratedInTheFutureWithinAllowedClockSkew
+        created_at_future = dt.datetime.now(tz=timezone.utc) + dt.timedelta(minutes=29)  #max allowed clock skew is 30m
+        for expected_scope, expected_version in test_cases_all_scopes_all_versions:
+            with self.subTest(expected_scope=expected_scope, expected_version=expected_version):
+                token = generate_uid_token(expected_scope, expected_version, expires_at=created_at_future)
+                mock_refresh_bidstream_keys.return_value = EncryptionKeysCollection([master_key, site_key],
+                                                                                  expected_scope, site_id, 1,
+                                                99999, 86400)
+                self._client.refresh_keys()
+                result = self._client.decrypt_token_into_raw_uid(token, None)
+                self.assertIsNotNone(result)
+                self.assertEqual(result.identity_scope, expected_scope)
+                self.assertEqual(result.advertising_token_version, expected_version)
+
+    def test_empty_keys(self, mock_refresh_bidstream_keys):  # EmptyKeyContainer
```

Comment:
where are the 3 legacy tests and `IdentityScopeAndType_TestCases`?

Reply:
Had a look at the test `LegacyResponseFromOldOperator`, functionally its same as the `SmokeTest` so decided not to add it. 
Will add the remaining two legacy tests

Reply:
Also added IdentityScopeAndType_TestCases

Reply:
It's different than `SmokeTest`: `LegacyResponseFromOldOperator` uses the JSON response expected from an old version of Operator, as this is what we expect from Private Operators that haven't been upgraded.

Reply:
understood now. Adding this test case